controllers/integrationpipeline/integrationpipeline_adapter.go

```diff
@@ -227,6 +228,42 @@ func (a *Adapter) EnsureStatusReportedInSnapshot() (controller.OperationResult,
 	return controller.ContinueProcessing()
 }
 
+// EnsureFinalizersAreCalled is an operation that will ensure that finalizers are called whenever the
+// Integration PipelineRun being processed is marked for deletion. Once finalizers get called,
+// the finalizer will be removed and the Integration PipelineRun will go back to the queue, so it gets deleted.
+// If a finalizer function fails its execution or a finalizer fails to be removed,
+// the Integration PipelineRun will be requeued with the error attached.
+func (a *Adapter) EnsureFinalizersAreCalled() (controller.OperationResult, error) {
+	// Check if the Integration PipelineRun is marked for deletion and continue processing other operations otherwise
+	if a.pipelineRun.GetDeletionTimestamp() == nil {
+		return controller.ContinueProcessing()
+	}
+
+	if controllerutil.ContainsFinalizer(a.pipelineRun, h.IntegrationPipelineRunFinalizer) {
+		status, _, err := GetIntegrationPipelineRunStatus(a.client, a.context, a.pipelineRun)
+		if err != nil {
+			return controller.RequeueWithError(err)
+		}
+		if status == gitops.IntegrationTestStatusInProgress {
+			return controller.ContinueProcessing()
+		}
+
+		patch := client.MergeFrom(a.pipelineRun.DeepCopy())
+		controllerutil.RemoveFinalizer(a.pipelineRun, h.IntegrationPipelineRunFinalizer)
+		err = a.client.Patch(a.context, a.pipelineRun, patch)
+		if err != nil {
+			return controller.RequeueWithError(err)
+		}
+
+		a.logger.Info("Removed Finalizer from the Integration PipelineRun",
```

Comment:
I think this should be logged like an audit event - like e.g. [here](https://github.com/redhat-appstudio/integration-service/blob/8f740239643c4a1f5f2f298118a915eae2cabbfe/controllers/integrationpipeline/integrationpipeline_adapter.go#L105-L106).

Reply:
Thanks! Updated it with `LogActionUpdate`, I hope it's the correct one.
For future reference, when should I log an audit event?

Reply:
We should do it whenever we change resources on the cluster - so when creating, updating, or deleting them.